Bug: `font-family-fallbacks` reports an error for generic fonts written with uppercase letters

### Environment

ESLint version: v10.8.0
@eslint/css version: v1.4.0
Node version: v22.22.3
npm version: v10.8.2
Operating System: StackBlitz WebContainer (linux)

### Which language are you using?

stylesheet

### What did you do?

```js
// Configuration
// eslint.config.js
import css from "@eslint/css";

export default [
	{
		files: ["**/*.css"],
		plugins: { css },
		language: "css/css",
		rules: { "css/font-family-fallbacks": "error" },
	},
];
```

```css
a { font-family: Arial, Sans-Serif; }
b { font-family: Arial, SERIF; }
c { font-family: Serif; }
d { font: 16px Arial, MONOSPACE; }

/* control cases, correctly not reported */
e { font-family: Arial, sans-serif; }
f { font-family: serif; }
```

### What did you expect to happen?

I expected no errors to be reported for any of the six declarations above.

Generic family keywords are pre-defined CSS keywords, and the CSS specification states that keywords are matched ASCII case-insensitively. From [CSS Values and Units Module Level 4, §4.1 Pre-defined Keywords](https://www.w3.org/TR/css-values-4/#keywords):

(Keywords are identifiers and are interpreted ASCII case-insensitively (i.e., [a-z] and [A-Z] are equivalent))

Generic families are keywords only when they are unquoted — [CSS Fonts Module Level 4](https://www.w3.org/TR/css-fonts-4/#generic-family-value) notes that `<generic-font-family>` keywords cannot be quoted, otherwise they are interpreted as a `<font-family-name>`. All four cases above are unquoted, so they are the keyword form.

So `Sans-Serif`, `SERIF` and `MONOSPACE` are valid generic font families and mean exactly the same thing as their lowercase forms. Only the lowercase forms are accepted right now.

### What actually happened?

The four declarations that use uppercase letters are reported, while the two lowercase control cases are not:

```
❯ npx eslint .

/home/projects/stackblitz-starters-adhqdqr2/test.css
  1:18  error  Use a generic font last                     css/font-family-fallbacks
  2:18  error  Use a generic font last                     css/font-family-fallbacks
  3:18  error  Use fallback fonts and a generic font last  css/font-family-fallbacks
  4:11  error  Use a generic font last                     css/font-family-fallbacks

✖ 4 problems (4 errors, 0 warnings)
```

The fact that only the uppercase variants are reported shows that the cause is case-sensitive matching.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-adhqdqr2

### Participation

- [x] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

Hello. I looked into where this comes from, and I would like to share what I found.

**Where the bug is**

In `src/rules/font-family-fallbacks.js`, the `genericFonts` set is defined entirely in lowercase (L20–L34), and the candidate tokens are compared against it without any case normalization:

| Line | Code |
| --- | --- |
| L101 | `!genericFonts.has(valueList.at(-1))` |
| L183 / L189 | `genericFonts.has(variableList[...])` |
| L198 | `genericFonts.has(valueArr[0].name)` |
| L261 | `genericFonts.has(fontsList.at(-1))` |
| L273 | `genericFonts.has(lastFont.name)` |
| L356 / L389 / L400 / L444 | `genericFonts.has(...)` |

What makes me think this is an oversight rather than an intentional decision is that the same file already normalizes case for the other kind of keyword it handles. CSS-wide keywords are lowercased in two places:

```js
// L43
return cssWideKeywords.has(value.trim().toLowerCase());

// L135, inside create()
sourceCode.lexer.cssWideKeywords.map(keyword => keyword.toLowerCase())
```

So CSS-wide keywords are treated case-insensitively, but generic font keywords in the very same rule are not.

**Precedent**

While searching for duplicates I found that this repository has fixed this same class of bug already:

- #222 `fix: make relative-font-units unit matching case-insensitive` 

`font-family-fallbacks` looks like the one rule that was left out. I could not find an existing issue or pull request covering this, so I am reporting it.

If this is accepted, I'd like to take it on, following the same approach as #222.

--- a/src/rules/font-family-fallbacks.js
+++ b/src/rules/font-family-fallbacks.js
@@ -33,6 +33,26 @@ const genericFonts = new Set([
 	"fangsong",
 ]);
 
+/*
+ * Matches a generic font family as a whole word, for the single value `font`
+ * branch that searches the shorthand where the family sits next to the size.
+ * A name such as `MySerifFont` must not match.
+ */
+const genericFontPattern = new RegExp(
+	String.raw`(?<![\w-])(?:${[...genericFonts].join("|")})(?![\w-])`,
+	"iu",
+);
+
+/**
+ * Check if the name is a generic font family.
+ * Generic font families are keywords, so they are matched case-insensitively.
+ * @param {string} name The name to check.
+ * @returns {boolean} True if the name is a generic font family, false otherwise.
+ */
+function isGenericFont(name) {
+	return genericFonts.has(name.toLowerCase());
+}
+
 /**
  * Check if the value is a CSS-wide keyword.
  * @param {string} value The value to check.
@@ -87,9 +107,9 @@ function reportFontWithoutFallbacksInFontProperty(
 	const valueList = fontPropertyValues.split(",").map(v => v.trim());
 
 	if (valueList.length === 1) {
-		const containsGenericFont = Array.from(genericFonts).some(font =>
-			valueList[0].includes(font),
-		);
+		// A quoted name is a family name, not a keyword
+		const unquoted = valueList[0].replace(/"[^"]*"|'[^']*'/gu, "");
+		const containsGenericFont = genericFontPattern.test(unquoted);
 
 		if (!containsGenericFont) {
 			context.report({
@@ -98,7 +118,7 @@ function reportFontWithoutFallbacksInFontProperty(
 			});
 		}
 	} else {
-		if (!genericFonts.has(valueList.at(-1))) {
+		if (!isGenericFont(valueList.at(-1))) {
 			context.report({
 				loc: node.loc,
 				messageId: "useGenericFont",
@@ -180,13 +200,13 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 
 						if (
 							variableList.length === 1 &&
-							!genericFonts.has(variableList[0])
+							!isGenericFont(variableList[0])
 						) {
 							context.report({
 								loc: node.loc,
 								messageId: "useFallbackFonts",
 							});
-						} else if (!genericFonts.has(variableList.at(-1))) {
+						} else if (!isGenericFont(variableList.at(-1))) {
 							context.report({
 								loc: node.loc,
 								messageId: "useGenericFont",
@@ -195,7 +215,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 					} else {
 						if (
 							valueArr[0].type === "Identifier" &&
-							genericFonts.has(valueArr[0].name)
+							isGenericFont(valueArr[0].name)
 						) {
 							return;
 						}
@@ -230,7 +250,10 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 
 						valueArr.forEach(child => {
 							if (child.type === "String") {
-								fontsList.push(child.value);
+								// Keep the quotes so it is not read as a keyword
+								fontsList.push(
+									sourceCode.getText(child).trim(),
+								);
 							}
 
 							if (child.type === "Identifier") {
@@ -258,7 +281,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 
 						if (
 							fontsList.length > 0 &&
-							!genericFonts.has(fontsList.at(-1))
+							!isGenericFont(fontsList.at(-1))
 						) {
 							context.report({
 								loc: node.loc,
@@ -270,7 +293,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 
 						if (!(
 							lastFont.type === "Identifier" &&
-							genericFonts.has(lastFont.name)
+							isGenericFont(lastFont.name)
 						)) {
 							context.report({
 								loc: node.loc,
@@ -352,9 +375,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 								);
 
 								if (!usingVar) {
-									if (
-										!genericFonts.has(afterOperator.at(-1))
-									) {
+									if (!isGenericFont(afterOperator.at(-1))) {
 										context.report({
 											loc: node.loc,
 											messageId: "useGenericFont",
@@ -386,9 +407,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 
 										if (
 											variableList.length > 0 &&
-											!genericFonts.has(
-												variableList.at(-1),
-											)
+											!isGenericFont(variableList.at(-1))
 										) {
 											context.report({
 												loc: node.loc,
@@ -397,9 +416,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 										}
 									} else {
 										if (
-											!genericFonts.has(
-												afterOperator.at(-1),
-											)
+											!isGenericFont(afterOperator.at(-1))
 										) {
 											context.report({
 												loc: node.loc,
@@ -441,7 +458,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 								);
 							} else {
 								if (
-									!genericFonts.has(
+									!isGenericFont(
 										sourceCode
 											.getText(valueArr.at(-1))
 											.trim(),

--- a/src/rules/font-family-fallbacks.js
+++ b/src/rules/font-family-fallbacks.js
@@ -251,9 +251,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 						valueArr.forEach(child => {
 							if (child.type === "String") {
 								// Keep the quotes so it is not read as a keyword
-								fontsList.push(
-									sourceCode.getText(child),
-								);
+								fontsList.push(sourceCode.getText(child));
 							}
 
 							if (child.type === "Identifier") {

--- a/src/rules/font-family-fallbacks.js
+++ b/src/rules/font-family-fallbacks.js
@@ -252,7 +252,7 @@ export default /** @satisfies {FontFamilyFallbacksRuleDefinition} */ ({
 							if (child.type === "String") {
 								// Keep the quotes so it is not read as a keyword
 								fontsList.push(
-									sourceCode.getText(child).trim(),
+									sourceCode.getText(child),
 								);
 							}
 

--- a/tests/rules/font-family-fallbacks.test.js
+++ b/tests/rules/font-family-fallbacks.test.js
@@ -56,6 +56,19 @@ ruleTester.run("font-family-fallbacks", rule, {
 				},
 			},
 		},
+		"a { font-family: Serif; }",
+		"a { font-family: SANS-SERIF; }",
+		"a { font-family: Arial, Sans-Serif; }",
+		"a { font-family: 'Arial', 'Segoe UI Emoji', SERIF; }",
+		"a { font: 16px MONOSPACE; }",
+		"a { font: 16px Arial, MONOSPACE; }",
+		"a { font: italic bold 1.2em 'Open Sans', Sans-Serif; }",
+		":root { --my-font: SANS-SERIF; } a { font-family: var(--my-font); }",
+		":root { --my-font: 'Arial'; } a { font-family: var(--my-font), Sans-Serif; }",
+		"a { font: var(--font-size) 'Open Sans', MONOSPACE; }",
+		":root { --my-font: Arial, SANS-SERIF; } a { font: 16px/1.5 Helvetica, var(--my-font); }",
+		"a { font: 16px/1.5 Arial, var(--x), SANS-SERIF; }",
+		"a { font: var(--font-weight) var(--font-size)/var(--line-height) MONOSPACE; }",
 	],
 	invalid: [
 		{
@@ -271,6 +284,126 @@ ruleTester.run("font-family-fallbacks", rule, {
 					column: 59,
 					endLine: 1,
 					endColumn: 132,
+				},
+			],
+		},
+		{
+			code: 'a { font-family: Arial, "serif"; }',
+			errors: [
+				{
+					messageId: "useGenericFont",
+					line: 1,
+					column: 18,
+					endLine: 1,
+					endColumn: 32,
+				},
+			],
+		},
+		{
+			code: 'a { font-family: Arial, "SERIF"; }',
+			errors: [
+				{
+					messageId: "useGenericFont",
+					line: 1,
+					column: 18,
+					endLine: 1,
+					endColumn: 32,
+				},
+			],
+		},
+		{
+			code: 'a { font: 16px "SERIF"; }',
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 23,
+				},
+			],
+		},
+		{
+			code: 'a { font: 16px "MySerifFont"; }',
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 29,
+				},
+			],
+		},
+		{
+			code: 'a { font: 16px "serif"; }',
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 23,
+				},
+			],
+		},
+		{
+			code: "a { font: 16px MySerifFont; }",
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 27,
+				},
+			],
+		},
+		{
+			code: "a { font: 16px MyserifFont; }",
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 27,
+				},
+			],
+		},
+		{
+			code: "a { font: 16px MYSERIFFONT; }",
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 27,
+				},
+			],
+		},
+		{
+			code: "a { font: 16px MyMonospaceFont; }",
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 31,
+				},
+			],
+		},
+		{
+			code: ":root { --my-font: 'Arial'; } a { font-family: var(--my-font), \"sans-serif\"; }",
+			errors: [
+				{
+					messageId: "useGenericFont",
+					line: 1,
+					column: 48,
+					endLine: 1,
+					endColumn: 76,
 				},
 			],
 		},

--- a/tests/rules/font-family-fallbacks.test.js
+++ b/tests/rules/font-family-fallbacks.test.js
@@ -407,5 +407,53 @@ ruleTester.run("font-family-fallbacks", rule, {
 				},
 			],
 		},
+		{
+			code: "a { font: 16px 'Roboto', 'Open Sans', 'serif'; }",
+			errors: [
+				{
+					messageId: "useGenericFont",
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 46,
+				},
+			],
+		},
+		{
+			code: ":root { --my-font: 'serif'; } a { font: 16px var(--my-font); }",
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 41,
+					endLine: 1,
+					endColumn: 60,
+				},
+			],
+		},
+		{
+			code: ":root { --my-font: 'MySerifFont'; } a { font: 16px var(--my-font); }",
+			errors: [
+				{
+					messageId: "useFallbackFonts",
+					line: 1,
+					column: 47,
+					endLine: 1,
+					endColumn: 66,
+				},
+			],
+		},
+		{
+			code: ":root { --my-font: 'Roboto', 'Open Sans', 'serif'; } a { font: 16px var(--my-font); }",
+			errors: [
+				{
+					messageId: "useGenericFont",
+					line: 1,
+					column: 64,
+					endLine: 1,
+					endColumn: 83,
+				},
+			],
+		},
 	],
 });

